### PR TITLE
Bug 1143247: Add support for validating the contents of arrow function e...

### PR DESCRIPTION
**Pull request by kmaglione**
...xpressions.

### CODE REVIEW

**Comment by diox** [comment]
This needs a test

**Comment by kmaglione** [comment]
Added tests.

**Comment by magopian** [comment]
@kmaglione could you please rebase?

File: tests/test_js_instanceactions.py [hunk 1]
```diff
@@ -101,19 +101,26 @@ def test_callexpression_argument_traversal():
...
-
-    assert _do_test_raw("""
-    function foo(x){}
-    foo({"bar":function(){
-        eval("evil");
-    }});
-    """).failed()
+    DECLARATIONS = (
+        "function foo(x){}",
+        "var foo = function foo(x){}",
```
Comment (magopian) [inline tests/test_js_instanceactions.py]: Seems this one was missing previously, good catch :+1:

File: validator/testcases/javascript/nodedefinitions.py [hunk 2]
```diff
@@ -57,6 +57,9 @@ def node(branches=None, dynamic=False, action=None, returns=False,
     "FunctionExpression": node(branches=("body", ), dynamic=True,
                                action=actions._func_expr, returns=True,
                                is_block=True),
+    "ArrowExpression": node(branches=("body", ), dynamic=True,
+                            action=actions._func_expr, returns=True,
+                            is_block=True),
```
Comment (magopian) [inline validator/testcases/javascript/nodedefinitions.py]: It's exactly the same as the `FunctionExpression`, what's its use?

File: tests/test_js_syntax.py [hunk 3]
```diff
@@ -56,3 +56,42 @@ def test_spidermonkey_warning():
...
+        "if (true) ; else { %s }",
+        "while (true) { %s }",
+        "do { %s } while (true)",
+        "for (;;) { %s }",
+        "try { %s } catch (e) {}",
+        "try {} catch (e) { %s }",
+        "try {} finally { %s }",
+        "try {} catch (e if %s) {}",
+        # This is broken.
+        # "let (foo = bar) { %s }",
```
Comment (magopian) [inline tests/test_js_syntax.py]: Is that a `TODO`? Do you want to create a new bug if it needs to get fixed, and add that bug number to the comment?

File: tests/test_js_syntax.py [hunk 4]
```diff
@@ -56,3 +56,42 @@ def test_spidermonkey_warning():
...
+    Tests that blocks of code are actually evaluated under normal
+    circumstances.
+    """
+
+    ID = ("js", "traverser", "dangerous_global")
+
+    EVIL = "eval(evilStuff)"
+    BLOCKS = (
+        "function foo() { %s; }",
+        "function foo() { %s; yield 1; }",
```
Comment (magopian) [inline tests/test_js_syntax.py]: What's the use of this `yield 1;` here (and below)?

Comment (magopian) [inline tests/test_js_syntax.py]: Ah, I see from the MDN docs that it's legacy generator functions, TIL ;)

**Comment by magopian** [comment]
r+wc

File: tests/test_js_syntax.py (hunk 3 above)
Comment (kmaglione) [inline tests/test_js_syntax.py]: It's https://bugzil.la/1009181, but support for `let` blocks has finally been removed, so I'll just remove the comment.

File: validator/testcases/javascript/nodedefinitions.py (hunk 2 above)
Comment (kmaglione) [inline validator/testcases/javascript/nodedefinitions.py]: For the purposes of the validator, it's exactly the same. There are some semantic differences, the main ones being:

* It captures the `this` object of its enclosing scope.
* It does not have access to the `arguments` object.

The main use, though, is that it's just shorter to type.

**Comment by kmaglione** [comment]
Rebased and removed outdated comment.

**Pull request merged**